models/stock_quant_package.py

```diff
@@ -79,3 +79,54 @@ def has_same_content(self, other):
         self.ensure_one()
         return frozenset(self._get_all_products_quantities().items()) == \
                frozenset(other._get_all_products_quantities().items())
+
+    def assert_reserved_full_package(self, picking):
+        """ Check that a package is fully reserved for a picking."""
+        MoveLine = self.env['stock.move.line']
+
+        self.ensure_one()
+
+        move_lines = picking.mapped('move_line_ids').filtered(lambda ml: ml.package_id == self)
+        pack_products = frozenset(self._get_all_products_quantities().items())
+        mls_products = frozenset(move_lines._get_all_products_quantities().items())
+        if pack_products != mls_products:
+            # move_lines do not match the quants
+            #picking = move_lines.mapped('picking_id')
+            #picking.ensure_one()
+            pack_mls = MoveLine.search([('package_id', 'child_of', self.id),
+                                        ('state', 'not in', ['done', 'cancel'])
+                                        ])
+            other_pickings = pack_mls.mapped('picking_id') - picking
+            if other_pickings:
+                raise ValidationError(
+                    _('The package is reserved in other pickings:') %
+                    ','.join(other_pickings.mapped('name'))
+                )
+            # other_pickings == False means partially reserved,
+            raise ValidationError(
+                _('Cannot mark as done a partially reserved package.')
+            )
+            """
+            # therefore reserve the remaining quantities
+            # TODO: this code can be used for whole package reservation
+            quants = self._get_contained_quants()
+            remaining_qtys = quants.group_quantity_by_product(only_available=True)
+            picking.with_context(
+                quant_ids=quants.ids)._create_moves(remaining_qtys,
+                                                   confirm=True,
+                                                   assign=True)
+            move_lines = picking.mapped('move_line_ids').filtered(lambda ml: ml.package_id == self)
+            """
+        #return move_lines
+
+    # TODO Fix in odoo core (copy pasted from there)
+    def _get_all_products_quantities(self):
+        '''This function computes the different product quantities for the given package
+        '''
+        # TDE CLEANME: probably to move somewhere else, like in pack op
+        res = {}
+        for quant in self._get_contained_quants():
+            if quant.product_id not in res:
```

Comment:
defaultdict again?

Reply:
this comes from odoo core, I'll leave it as it is, since this needs to be changed there